src/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs

```diff
@@ -0,0 +1,327 @@
+
+                // Determine how many waiters to wake, taking into account how many spinners and waiters there are and how many waiters
+                // have previously been signaled to wake but have not yet woken
+                countOfWaitersToWake =
+                    (int)Math.Min(newCounts._signalCount, (uint)newCounts._waiterCount + newCounts._spinnerCount) -
+                    newCounts._spinnerCount -
+                    newCounts._countOfWaitersSignaledToWake;
+                if (countOfWaitersToWake > 0)
+                {
+                    // Ideally, limiting to a maximum of releaseCount would not be necessary and could be an assert instead, but since
+                    // WaitForSignal() does not have enough information to tell whether a woken thread was signaled, and due to the cap
+                    // below, it's possible for countOfWaitersSignaledToWake to be less than the number of threads that have actually
+                    // been signaled to wake.
+                    if (countOfWaitersToWake > releaseCount)
+                    {
+                        countOfWaitersToWake = releaseCount;
+                    }
+
+                    // Cap countOfWaitersSignaledToWake to its max value. It's ok to ignore some woken threads in this count, it just
+                    // means some more threads will be woken next time. Typically, it won't reach the max anyway.
+                    newCounts._countOfWaitersSignaledToWake += (byte)Math.Min(countOfWaitersToWake, byte.MaxValue);
+                    if (newCounts._countOfWaitersSignaledToWake <= counts._countOfWaitersSignaledToWake)
+                    {
+                        newCounts._countOfWaitersSignaledToWake = byte.MaxValue;
+                    }
+                }
+
+                Counts countsBeforeUpdate = Counts.CompareExchange(ref _separated._counts, newCounts, counts);
+                if (countsBeforeUpdate == counts)
+                {
+                    Debug.Assert(releaseCount <= _maximumSignalCount - counts._signalCount);
+                    if (countOfWaitersToWake > 0)
+                        ReleaseCore(countOfWaitersToWake);
+                    return;
+                }
+
+                counts = countsBeforeUpdate;
+            }
+        }
+
+        private bool WaitForSignal(int timeoutMs)
+        {
+            Debug.Assert(timeoutMs > 0);
+
+            while (true)
+            {
+                if (!WaitCore(timeoutMs))
+                {
+                    // Unregister the waiter. The wait subsystem used above guarantees that a thread that wakes due to a timeout does
+                    // not observe a signal to the object being waited upon.
+                    Counts toSubtract = new Counts();
+                    toSubtract._waiterCount++;
+                    Counts newCounts = Counts.Subtract(ref _separated._counts, toSubtract);
+                    Debug.Assert(newCounts._waiterCount >= 0);
```

Comment:
Since `_waiterCount` is unsigned, I believe this condition is never false. Could instead assert that `(ushort)(newCounts._waiterCount + 1) != 0`.